primitives/npos-elections/fuzzer/src/phragmen_pjr.rs

```diff
@@ -0,0 +1,88 @@
+// This file is part of Substrate.
+
+// Copyright (C) 2020-2021 Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: Apache-2.0
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Fuzzing which ensures that running unbalanced sequential phragmen always produces a result
+//! which satisfies our PJR checker.
+//!
+//! # Running
+//!
+//! Run with `cargo hfuzz run phragmen_pjr`.
+//!
+//! # Debugging a panic
+//!
+//! Once a panic is found, it can be debugged with
+//! `cargo hfuzz run-debug phragmen_pjr hfuzz_workspace/phragmen_pjr/*.fuzz`.
+
+use honggfuzz::fuzz;
+
+mod common;
+use common::{generate_random_npos_inputs, to_range};
+use rand::{self, SeedableRng};
+use sp_npos_elections::{
+	pjr_check, seq_phragmen, to_supports, ElectionResult, ExtendedBalance, StakedAssignment,
+	Supports,
+};
+
+type AccountId = u64;
+type PerThing = sp_arithmetic::PerU16;
+
+/// TODO: This value is currently _entirely arbitrary_.
+const PJR_THRESHOLD: ExtendedBalance = 1_000_000_000_000;
+
+fn main() {
+	loop {
+		fuzz!(|data: (usize, usize, u64)| {
+			let (mut candidate_count, mut voter_count, seed) = data;
+			let rng = rand::rngs::SmallRng::seed_from_u64(seed);
+			candidate_count = to_range(candidate_count, 100, 1000);
+			voter_count = to_range(voter_count, 100, 2000);
+
+			let (rounds, candidates, voters) =
+				generate_random_npos_inputs(candidate_count, voter_count, rng);
+
+			// Run the NPoS validator selection algorithm
+			let ElectionResult {
+				winners,
+				assignments,
+			} = seq_phragmen::<AccountId, PerThing>(rounds, candidates.clone(), voters.clone(), None)
+				.expect("seq_phragmen must succeed");
+
+			// pjr_check only cares about the identity of the winner, not its balance
+			let winners: Vec<_> = winners.into_iter().map(|(id, _stake)| id).collect();
+
+			// convert assignments into staked assignments
+			let assignments: Vec<StakedAssignment<AccountId>> = assignments
```

Comment:
there's a method for this in the helpers, if you can build a closure that gives the weight.

Reply:
Good idea: fc64ff572